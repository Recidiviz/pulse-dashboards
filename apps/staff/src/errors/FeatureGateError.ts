// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

/**
 * Exception to be thrown when a feature flag forbids progress.
 * Useful in situations where the flag cannot be checked until after data is fetched
 * and you need to abort hydration when the flag is turned off.
 */
export class FeatureGateError extends Error {}
