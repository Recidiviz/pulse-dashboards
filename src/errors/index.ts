export { FeatureGateError } from "./FeatureGateError";
export { OpportunityValidationError } from "./OpportunityValidationError";
export { TaskValidationError } from "./TaskValidationError";
