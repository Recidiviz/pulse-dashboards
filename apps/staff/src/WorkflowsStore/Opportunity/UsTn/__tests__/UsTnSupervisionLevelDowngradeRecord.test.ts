// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { identity } from "lodash";

import { OpportunityValidationError } from "../../../../errors";
import {
  getSLDValidator as getValidator,
  UsTnSupervisionLevelDowngradeReferralRecordRaw,
  usTnSupervisionLevelDowngradeReferralRecordSchemaForSupervisionLevelFormatter,
} from "../UsTnSupervisionLevelDowngradeOpportunity/UsTnSupervisionLevelDowngradeReferralRecord";

const usTnSupervisionLevelDowngradeRecordRaw: UsTnSupervisionLevelDowngradeReferralRecordRaw =
  {
    stateCode: "US_XX",
    externalId: "abc123",
    eligibleCriteria: {
      supervisionLevelHigherThanAssessmentLevel: {
        assessmentLevel: "HIGH",
        latestAssessmentDate: "2021-08-20",
        supervisionLevel: "MAXIMUM",
      },
    },
    ineligibleCriteria: {},
    caseNotes: {
      Violations: [
        { eventDate: "2021-05-03", noteBody: "VRPT" },
        { eventDate: "2020-01-10", noteBody: "ARRP" },
      ],
    },
  };

const mockClient = {
  supervisionLevel: "MAXIMUM",
  rootStore: {
    workflowsStore: {
      formatSupervisionLevel: identity,
    },
  },
};

const getTransformedRecord = () =>
  usTnSupervisionLevelDowngradeReferralRecordSchemaForSupervisionLevelFormatter(
    identity,
  ).parse(usTnSupervisionLevelDowngradeRecordRaw);

test("transform function", () => {
  expect(getTransformedRecord()).toMatchSnapshot();
});

test("record validates", () => {
  const validator = getValidator(mockClient as any);
  expect(() => validator(getTransformedRecord())).not.toThrow();
});

test("record does not validate", () => {
  const validator = getValidator({
    ...mockClient,
    supervisionLevel: "MEDIUM",
  } as any);
  expect(() => validator(getTransformedRecord())).toThrow(
    OpportunityValidationError,
  );
});
