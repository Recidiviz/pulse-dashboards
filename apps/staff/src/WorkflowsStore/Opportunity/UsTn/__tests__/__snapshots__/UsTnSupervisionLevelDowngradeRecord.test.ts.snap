// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform function 1`] = `
{
  "caseNotes": {
    "Violations": [
      {
        "eventDate": 2021-05-03T00:00:00.000Z,
        "noteBody": "VRPT",
      },
      {
        "eventDate": 2020-01-10T00:00:00.000Z,
        "noteBody": "ARRP",
      },
    ],
  },
  "eligibleCriteria": {
    "supervisionLevelHigherThanAssessmentLevel": {
      "assessmentLevel": "High",
      "latestAssessmentDate": 2021-08-20T00:00:00.000Z,
      "supervisionLevel": "MAXIMUM",
    },
  },
  "externalId": "abc123",
  "ineligibleCriteria": {},
  "stateCode": "US_XX",
}
`;
